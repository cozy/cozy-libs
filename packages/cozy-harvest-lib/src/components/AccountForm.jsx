import React, { PureComponent } from 'react'
import { Form, Field as FinalFormField } from 'react-final-form'
import PropTypes from 'react-proptypes'

import Button from 'cozy-ui/react/Button'
import { translate, extend } from 'cozy-ui/react/I18n'
import Field from 'cozy-ui/react/Field'

import AccountFormError from './AccountFormError'
import {
  getEncryptedFieldName,
  getFieldPlaceholder,
  sanitizeSelectProps
} from '../helpers/fields'
import { KonnectorJobError } from '../helpers/konnectors'
import manifest from 'helpers/manifest'
import OAuthForm from './OAuthForm'

const predefinedLabels = [
  'answer',
  'birthdate',
  'code',
  'date',
  'email',
  'firstname',
  'lastname',
  'login',
  'password',
  'phone'
]

// Out of scope labels already used, should be transferred directly in manifests
// in the future.
const legacyLabels = [
  'branchName' // Used in banking konnectors
]

const VALIDATION_ERROR_REQUIRED_FIELD = 'VALIDATION_ERROR_REQUIRED_FIELD'

export class AccountField extends PureComponent {
  render() {
    const {
      disabled,
      hasError,
      initialValue,
      label,
      name,
      onKeyUp,
      required,
      role,
      t,
      type
    } = this.props

    // Allow manifest to specify predefined label
    const localeKey =
      predefinedLabels.includes(label) || legacyLabels.includes(label)
        ? label
        : name

    const isEditable = !(role === 'identifier' && initialValue)

    const fieldProps = {
      ...this.props,
      autoComplete: 'off',
      className: 'u-m-0', // 0 margin
      disabled: disabled || !isEditable,
      error: hasError,
      fullwidth: true,
      label: t(`fields.${localeKey}.label`, {
        _: t(`legacy.fields.${localeKey}.label`, { _: name })
      }),
      onKeyUp: onKeyUp,
      placeholder: getFieldPlaceholder(
        this.props,
        t(`fields.${name}.placeholder`, { _: '' })
      ),
      side: !required && t('accountForm.fields.optional'),
      size: 'medium'
    }
    const passwordLabels = {
      hideLabel: t('accountForm.password.hide'),
      showLabel: t('accountForm.password.show')
    }
    switch (type) {
      case 'date':
        return (
          <Field
            {...fieldProps}
            placeholder={t(`fields.${name}.placeholder`, {
              _: t(`default.dateFormat`).toLowerCase()
            })}
          />
        )
      case 'dropdown':
        return <Field {...sanitizeSelectProps(fieldProps)} />
      case 'password':
        return (
          <Field
            {...fieldProps}
            // Usting the `new-password` value is the best way to avoid
            // autocomplete for password.
            // See https://stackoverflow.com/a/17721462/1135990
            autoComplete="new-password"
            secondaryLabels={passwordLabels}
          />
        )
      default:
        return <Field {...fieldProps} type="text" />
    }
  }
}

AccountField.propTypes = {
  hasError: PropTypes.bool,
  initialValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  onKeyUp: PropTypes.func,
  role: PropTypes.string,
  type: PropTypes.oneOf(['date', 'dropdown', 'email', 'password', 'text']),
  t: PropTypes.func
}

// As SelectBox component from Cozy-UI, rendering dropdown type, is just giving
// us the full Option object, we just get its value to facilitate mapping
// with account
const parse = type => value => {
  return type === 'dropdown' ? value.value : value
}

export class AccountFields extends PureComponent {
  render() {
    const {
      container,
      disabled,
      hasError,
      initialValues,
      manifestFields,
      onKeyUp,
      t
    } = this.props

    // Ready to use named fields array
    const namedFields = Object.keys(manifestFields).map(fieldName => ({
      ...manifestFields[fieldName],
      name: fieldName
    }))

    return (
      <div>
        {namedFields.map((field, index) => (
          <FinalFormField
            key={index}
            name={field.name}
            parse={parse(field.type)}
          >
            {({ input }) => (
              <AccountField
                {...field}
                {...input}
                container={container}
                hasError={hasError}
                disabled={disabled}
                initialValue={
                  initialValues[field.name] ||
                  initialValues[getEncryptedFieldName(field.name)]
                }
                onKeyUp={onKeyUp}
                t={t}
              />
            )}
          </FinalFormField>
        ))}
      </div>
    )
  }
}

AccountFields.propTypes = {
  disabled: PropTypes.bool,
  hasError: PropTypes.bool,
  fillEncrypted: PropTypes.bool,
  manifestFields: PropTypes.object.isRequired,
  onKeyUp: PropTypes.func,
  t: PropTypes.func
}

export class AccountForm extends PureComponent {
  constructor(props, context) {
    super(props, context)
    const { konnector, lang } = props
    const { locales } = konnector
    if (locales && lang) {
      extend(locales[lang])
    }
  }

  /**
   * Indicates if the state of ReactFinalForm implies that data can be submitted
   * @param  {Object} formState See https://github.com/final-form/final-form#formstate
   * @return {Boolean}
   */
  isSubmittable({ dirty, error, initialValues, valid }) {
    const untouched = initialValues && !dirty
    return error || (valid && !untouched)
  }

  handleKeyUp(event, { dirty, form, initialValues, valid, values }) {
    if (
      event.code === 'Enter' &&
      this.isSubmittable({ dirty, initialValues, valid })
    ) {
      this.handleSubmit(values, form)
    }
  }

  handleSubmit(values, form) {
    const { onSubmit } = this.props
    // Reset form with new values to set back dirty to false
    form.reset(values)
    onSubmit(values)
  }

  validate = (fields, initialValues) => vals => {
    let errors = {}
    for (let name in fields)
      if (
        fields[name].required &&
        !vals[name] &&
        // Don't require value for empty encrypted fields with initial value
        !initialValues[getEncryptedFieldName(name)]
      )
        errors[name] = VALIDATION_ERROR_REQUIRED_FIELD
    return errors
  }

  render() {
    const { account, error, konnector, onSubmit, submitting, t } = this.props
    const { fields, oauth } = konnector

    if (oauth) return <OAuthForm initialValues={initialValues} oauth={oauth} />

    const sanitizedFields = manifest.sanitizeFields(fields)
    const defaultValues = manifest.defaultFieldsValues(sanitizedFields)
    const initialValues = account && account.auth
    const initialAndDefaultValues = { ...defaultValues, ...initialValues }

    let container = null

    return (
      <Form
        initialValues={initialAndDefaultValues}
        onSubmit={onSubmit}
        validate={this.validate(sanitizedFields, initialAndDefaultValues)}
        render={({ dirty, form, values, valid }) => (
          <div
            ref={element => {
              container = element
            }}
          >
            {error && (
              <AccountFormError error={error} konnector={konnector} t={t} />
            )}
            <AccountFields
              container={container}
              disabled={submitting}
              hasError={
                error &&
                error instanceof KonnectorJobError &&
                error.isLoginError()
              }
              initialValues={initialAndDefaultValues}
              manifestFields={sanitizedFields}
              onKeyUp={event =>
                this.handleKeyUp(event, {
                  dirty,
                  form,
                  initialValues,
                  valid,
                  values
                })
              }
              t={t}
            />
            <Button
              busy={submitting}
              className="u-mt-2 u-mb-1-half"
              disabled={
                submitting ||
                !this.isSubmittable({ dirty, error, initialValues, valid })
              }
              extension="full"
              label={t('accountForm.submit.label')}
              onClick={() => this.handleSubmit(values, form)}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  account: PropTypes.object,
  konnector: PropTypes.object.isRequired,
  error: PropTypes.object,
  submitting: PropTypes.bool
}

export default translate()(AccountForm)
