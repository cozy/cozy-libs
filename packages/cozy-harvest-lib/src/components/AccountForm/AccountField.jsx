import Field from 'cozy-ui/transpiled/react/Field'
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { translate } from 'cozy-ui/transpiled/react/I18n'

import { getFieldPlaceholder, sanitizeSelectProps } from '../../helpers/fields'
import {
  legacyLabels,
  predefinedLabels,
  ROLE_IDENTIFIER
} from '../../helpers/manifest'
import { isIOS } from 'cozy-device-helper'

// On iOS, focusing an input not in the same tick as the user action is buggy.
// The field is focused but the keyboard does not show up,
// See https://github.com/apache/cordova-plugin-wkwebview-engine/pull/37/#issuecomment-308321094
const canAutofocus = () => {
  if (isIOS()) {
    return false
  }
  return true
}

/**
 * AccountField encapsulate an unique Cozy-UI Field component.
 * It maps its own props to Cozy-UI Field ones, and deal with i18n and locale
 * logic.
 *
 * AccountField also expect to receive all properties defined in manifest
 * fields.
 */
export class AccountField extends PureComponent {
  constructor(props) {
    super(props)
    // Ref to identifier input, needed to give focus to identifier
    this.inputRef = null
    this.setInputRef = this.setInputRef.bind(this)
  }

  componentDidMount() {
    const { role } = this.props
    if (role === ROLE_IDENTIFIER && this.inputRef && canAutofocus()) {
      this.inputRef.focus()
    }
  }

  setInputRef(element) {
    this.inputRef = element

    const { inputRef } = this.props
    if (typeof inputRef === 'function') {
      inputRef(element)
    }
  }

  render() {
    const {
      disabled,
      hasError,
      forceEncryptedPlaceholder,
      label,
      name,
      required,
      t,
      type
    } = this.props

    // Allow manifest to specify predefined label
    const localeKey =
      predefinedLabels.includes(label) || legacyLabels.includes(label)
        ? label
        : name

    // Cozy-UI <Field /> props
    const fieldProps = {
      ...this.props,
      autoComplete: 'off',
      className: 'u-m-0', // 0 margin
      disabled: disabled,
      error: !disabled && hasError,
      fullwidth: true,
      inputRef: this.setInputRef,
      label: t(`fields.${localeKey}.label`, {
        _: t(`legacy.fields.${localeKey}.label`, { _: name })
      }),
      placeholder: getFieldPlaceholder(
        this.props,
        t(`fields.${name}.placeholder`, { _: '' }),
        { forceEncryptedPlaceholder }
      ),
      side: required ? null : t('accountForm.fields.optional'),
      size: 'medium'
    }

    const passwordLabels = {
      hideLabel: t('accountForm.password.hide'),
      showLabel: t('accountForm.password.show')
    }

    switch (type) {
      case 'date':
        return (
          <Field
            {...fieldProps}
            placeholder={t(`fields.${name}.placeholder`, {
              _: t(`default.dateFormat`).toLowerCase()
            })}
          />
        )
      case 'dropdown':
        return <Field {...sanitizeSelectProps(fieldProps)} />
      case 'password':
        return (
          /*
          Using the `new-password` value is the best way to avoid
          autocomplete for password.
          See https://stackoverflow.com/a/17721462/1135990
           */
          <Field
            {...fieldProps}
            autoComplete="new-password"
            secondaryLabels={passwordLabels}
          />
        )
      default:
        return <Field {...fieldProps} type="text" />
    }
  }
}

AccountField.propTypes = {
  /**
   * The element wrapping the <Field /> component.
   * Passed to <SelectBox /> component.
   */
  container: PropTypes.instanceOf(Element),
  /**
   * Indicates if the <Field /> should be rendered with an error style.
   */
  hasError: PropTypes.bool,
  /**
   * Initial value of the field
   */
  initialValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * A callback to call on Field input ref.
   */
  inputRef: PropTypes.func,
  /**
   * Optionnal predefined label, used as locale key.
   */
  label: PropTypes.oneOf([...predefinedLabels, ...legacyLabels]),
  /**
   * field name as declared in manifest, used as locale key and as fallback for
   * label
   */
  name: PropTypes.string.isRequired,
  /**
   * Indicates if the field is required. Non-required fields will be indicated
   * as "optional"
   */
  required: PropTypes.bool,
  /**
   * Indicates fields role. Defined in manifest or during fields sanitization.
   * Valid values are ['identifier', 'password'].
   * The field having role identifier will be given focus.
   */
  role: PropTypes.string,
  /**
   * Field type, passed to <Field /> component. Except for "dropdown" which
   * will be mapped to "select"
   */
  type: PropTypes.oneOf(['date', 'dropdown', 'email', 'password', 'text']),
  /**
   * Translation function
   */
  t: PropTypes.func
}

export default translate()(AccountField)
