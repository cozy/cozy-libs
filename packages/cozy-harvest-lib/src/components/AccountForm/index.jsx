import React, { PureComponent } from 'react'
import { Form } from 'react-final-form'
import PropTypes from 'prop-types'

import { isMobile } from 'cozy-device-helper'
import Button from 'cozy-ui/transpiled/react/Button'
import { translate } from 'cozy-ui/transpiled/react/I18n'

import AccountFields from './AccountFields'
import TriggerErrorInfo from '../infos/TriggerErrorInfo'
import { getEncryptedFieldName } from '../../helpers/fields'
import { KonnectorJobError } from '../../helpers/konnectors'
import manifest from '../../helpers/manifest'
import withKonnectorLocales from '../hoc/withKonnectorLocales'

const VALIDATION_ERROR_REQUIRED_FIELD = 'VALIDATION_ERROR_REQUIRED_FIELD'

/**
 * AccountForm is reponsible of generating a form which will allow user to
 * edit an account for a given konnector.
 *
 * This component rely internally on a ReactFinalForm component.
 *
 * AccountForm must be passed `onSubmit` callback.
 * @type {PureComponent}
 * @see https://github.com/final-form/react-final-form#getting-started
 */
export class AccountForm extends PureComponent {
  constructor(props) {
    super(props)

    this.inputs = {}
    this.inputFocused = null

    this.inputRefByName = this.inputRefByName.bind(this)
    this.handleFocus = this.handleFocus.bind(this)
    this.handleKeyUp = this.handleKeyUp.bind(this)
    this.focusNext = this.focusNext.bind(this)
  }

  /**
   * Indicates if the state of ReactFinalForm implies that data can be submitted
   * @param  {Object} formState See https://github.com/final-form/final-form#formstate
   * @return {Boolean}
   */
  isSubmittable({ dirty, error, initialValues, valid }) {
    const modified = !initialValues || dirty
    // Here error is not a validation error but an instance of a
    // KonnectorJobError, so submitting again is possible
    return modified && (error || valid)
  }

  /**
   * Give focus to next input in the form.
   * Fallback for mobile devices as we are using a div element instead of a
   * form.
   * @return {Element} Focused element or null if no element has been focused
   */
  focusNext() {
    if (!this.inputs) return null

    const inputs = Object.values(this.inputs)

    const currentIndex = inputs.indexOf(this.inputFocused)

    let nextIndex = currentIndex + 1
    let nextInput = inputs[nextIndex]

    if (nextInput) {
      nextInput.focus()
    }

    return nextInput || null
  }

  /**
   * Capture input with focus and store the current focused input
   * @param  {Event} event Focus event
   */
  handleFocus(event) {
    if (Object.values(this.inputs).includes(event.target)) {
      this.inputFocused = event.target
    }
  }

  /**
   * Handle key up and check if `ENTER` key has been pressed. If so, submit the
   * form if all parmater are ok.
   * @param  {React.SyntheticEvent} event         Key events
   * @param  {Boolean} dirty         Indicates if form is dirty, i.e. if values
   * have changed
   * @param  {Object} form          The form object injected by ReactFinalForm.
   * @param  {Object} initialValues Initial data injected into AccountForm
   * @param  {Boolean} valid        Indicates if the form data is valid
   * @param  {Object} values        Actual form values data
   */
  handleKeyUp(event, { dirty, form, initialValues, valid, values }) {
    if (event.key === 'Enter') {
      const changedFocus = isMobile() && !!this.focusNext()
      if (
        !changedFocus &&
        this.isSubmittable({ dirty, initialValues, valid })
      ) {
        this.handleSubmit(values, form)
      }
    }
  }

  /**
   * Submit handler
   * @param  {Object} values        Actual form values data
   * @param  {Object} form          The form object injected by ReactFinalForm.
   */
  handleSubmit(values, form) {
    const { onSubmit } = this.props
    // Reset form with new values to set back dirty to false
    form.reset(values)
    onSubmit(values)
  }

  /**
   * Callback passed to `<AccountFields />` element. Called with a field name,
   * it return a callback wich take an input ref as argument. It then stores
   * the input ref into AccountForm's inner `inputs` property.
   * With this method we keep a list of inputs existing in the form.
   * The inputs inner property is indexed with names to avoid duplicates in case
   * of re-render of input children.
   * @param  {string} fieldName Field name
   * @return {function}         Callback which take an input ref as argument.
   */
  inputRefByName(fieldName) {
    return input => {
      this.inputs[fieldName] = input
    }
  }

  validate = (fields, initialValues) => vals => {
    let errors = {}
    for (let name in fields)
      if (
        fields[name].required &&
        !vals[name] &&
        // Don't require value for empty encrypted fields with initial value
        !initialValues[getEncryptedFieldName(name)]
      )
        errors[name] = VALIDATION_ERROR_REQUIRED_FIELD
    return errors
  }

  render() {
    const {
      account,
      error,
      konnector,
      onSubmit,
      showError,
      submitting,
      t
    } = this.props
    const { fields } = konnector

    const sanitizedFields = manifest.sanitizeFields(fields)
    const defaultValues = manifest.defaultFieldsValues(sanitizedFields)
    const initialValues = account && account.auth
    const initialAndDefaultValues = { ...defaultValues, ...initialValues }

    let container = null
    console.log('error AccountFrom', error)
    const isLoginError =
      error instanceof KonnectorJobError && error.isLoginError()

    return (
      // See https://github.com/final-form/react-final-form#getting-started
      <Form
        initialValues={initialAndDefaultValues}
        onSubmit={onSubmit}
        validate={this.validate(sanitizedFields, initialAndDefaultValues)}
        render={({ dirty, form, values, valid }) => (
          <div
            onKeyUp={event =>
              this.handleKeyUp(event, {
                dirty,
                form,
                initialValues,
                valid,
                values
              })
            }
            onFocusCapture={this.handleFocus}
            ref={element => {
              container = element
            }}
          >
            {error && showError && (
              <TriggerErrorInfo
                className="u-mb-1"
                error={error}
                konnector={konnector}
              />
            )}
            <AccountFields
              container={container}
              disabled={submitting}
              fields={sanitizedFields}
              hasError={error && isLoginError}
              initialValues={initialAndDefaultValues}
              inputRefByName={this.inputRefByName}
              t={t}
            />
            <Button
              busy={submitting}
              className="u-mt-2 u-mb-1-half"
              disabled={
                submitting ||
                !this.isSubmittable({ dirty, error, initialValues, valid })
              }
              extension="full"
              label={t('accountForm.submit.label')}
              onClick={() => this.handleSubmit(values, form)}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  /**
   * Account document. Used to get intial form values.
   * If no account is passed, AccountForm will use empty initial values.
   * @type {Object}
   */
  account: PropTypes.object,
  /**
   * Existing error
   * @type {Error,KonnectorJobError}
   */
  error: PropTypes.object,
  /**
   * Konnector document. AccountForm will check the `fields` object to compute
   * fields.
   * @type {Object}
   */
  konnector: PropTypes.object.isRequired,
  /**
   * Submit callback
   * @type {Function}
   */
  onSubmit: PropTypes.func.isRequired,
  /**
   * Indicates if the AccountForm has to show errors. Sometimes errors may be
   * displayed elsewhere. However, a KonnectorJobError corresponding to a login
   * error is always displayed.
   * @type {Boolean}
   */
  showError: PropTypes.bool,
  /**
   * Indicates if the form should be rendered as submitting data or busy.
   * Typically updated after an `onSubmit` call.
   * @type {Object}
   */
  submitting: PropTypes.bool,
  /**
   * Translation function
   */
  t: PropTypes.func
}

AccountForm.defaultProps = {
  showError: true
}

export default translate()(withKonnectorLocales(AccountForm))
