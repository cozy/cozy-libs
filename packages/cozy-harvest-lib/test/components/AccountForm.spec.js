/* eslint-env jest */
import React from 'react'
import PropTypes from 'prop-types'
import { shallow, mount } from 'enzyme'

import { isMobile } from 'cozy-device-helper'

import { AccountForm } from 'components/AccountForm'

const fixtures = {
  konnector: {
    fields: {
      username: {
        type: 'text'
      },
      passphrase: {
        type: 'password'
      }
    }
  },
  konnectorWithOptionalFields: {
    fields: {
      test: {
        required: false,
        type: 'text'
      }
    }
  },
  account: {
    auth: {
      username: 'Toto',
      credentials_encrypted:
        'bmFjbGj8JQfzzfTQ2aGKTpI+HI9N8xKAQqPTPD6/84x5GyiHm2hdn7N6rO8cLTCnkdsnd2eFWJRf'
    }
  }
}

const onSubmit = jest.fn()
const t = jest.fn()

jest.mock('cozy-device-helper', () => ({
  ...require.requireActual('cozy-device-helper'),
  isMobile: jest.fn()
}))

jest.mock('components/KonnectorIcon', () => {
  const KonnectorIcon = () => <div>KonnectorIcon</div>

  return KonnectorIcon
})

describe('AccountForm', () => {
  beforeEach(() => {
    t.mockClear()
    t.mockImplementation(key => key)
    onSubmit.mockClear()
  })

  it('should render', () => {
    const wrapper = shallow(
      <AccountForm konnector={fixtures.konnector} onSubmit={onSubmit} t={t} />
    )
    const component = wrapper.dive().getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render error', () => {
    const wrapper = shallow(
      <AccountForm
        error={new Error('Test error')}
        konnector={fixtures.konnector}
        onSubmit={onSubmit}
        t={t}
      />
    )
    const component = wrapper.dive().getElement()
    expect(component).toMatchSnapshot()
  })

  it('should not render error', () => {
    const wrapper = shallow(
      <AccountForm
        error={new Error('Test error')}
        konnector={fixtures.konnector}
        onSubmit={onSubmit}
        showError={false}
        t={t}
      />
    )
    const component = wrapper.dive().getElement()
    expect(component).toMatchSnapshot()
  })

  it('should inject initial values from account', () => {
    const wrapper = shallow(
      <AccountForm
        account={fixtures.account}
        konnector={fixtures.konnector}
        onSubmit={onSubmit}
        t={t}
      />
    )
    expect(wrapper.props().initialValues).toEqual(fixtures.account.auth)
  })

  it('should provide default values from manifest', () => {
    const konnector = {
      fields: {
        foo: {
          default: 'bar',
          type: 'text'
        }
      }
    }
    const wrapper = shallow(
      <AccountForm konnector={konnector} onSubmit={onSubmit} t={t} />
    )
    expect(wrapper.props().initialValues).toEqual({
      foo: 'bar'
    })
  })

  describe('Submit Button', () => {
    const getButtonDisabledValue = wrapper =>
      wrapper
        .dive()
        .find('DefaultButton')
        .props().disabled

    const assertButtonDisabled = wrapper =>
      expect(getButtonDisabledValue(wrapper)).toBe(true)

    const assertButtonEnabled = wrapper =>
      expect(getButtonDisabledValue(wrapper)).toBe(false)

    it('should be disabled if there is required field empty', () => {
      const konnector = {
        fields: {
          test: {
            type: 'text'
          }
        }
      }

      assertButtonDisabled(
        shallow(
          <AccountForm konnector={konnector} onSubmit={onSubmit} t={t} />,
          // Avoid componentDidMount
          { disableLifecycleMethods: true }
        )
      )
    })

    it("should be enabled if required field isn't empty", () => {
      const konnector = {
        fields: {
          test: {
            default: 'test',
            type: 'text'
          }
        }
      }
      assertButtonEnabled(
        shallow(
          <AccountForm konnector={konnector} onSubmit={onSubmit} t={t} />,
          // Avoid componentDidMount
          { disableLifecycleMethods: true }
        )
      )
    })

    it("should be enabled if fields isn't required", () => {
      assertButtonEnabled(
        shallow(
          <AccountForm
            konnector={fixtures.konnectorWithOptionalFields}
            onSubmit={onSubmit}
            t={t}
          />,
          // Avoid componentDidMount
          { disableLifecycleMethods: true }
        )
      )
    })

    it('should be disabled with initialValues', () => {
      const account = {
        auth: {
          username: 'foo',
          passphrase: 'bar'
        }
      }

      assertButtonDisabled(
        shallow(
          <AccountForm
            account={account}
            konnector={fixtures.konnector}
            onSubmit={onSubmit}
            t={t}
          />,
          // Avoid componentDidMount
          { disableLifecycleMethods: true }
        )
      )
    })

    /**
     * This case corresponds to an account edition of a trigger already having
     * an error. We are not supposed to submit it without modifications.
     */
    it('should be disabled with initialValues AND error', () => {
      const account = {
        auth: {
          username: 'foo',
          passphrase: 'bar'
        }
      }

      assertButtonDisabled(
        shallow(
          <AccountForm
            account={account}
            error={new Error('Existing trigger error')}
            konnector={fixtures.konnector}
            onSubmit={onSubmit}
            t={t}
          />,
          // Avoid componentDidMount
          { disableLifecycleMethods: true }
        )
      )
    })

    it('should be enabled when an error exists', () => {
      const values = {
        username: 'foo',
        passphrase: 'bar'
      }
      assertButtonEnabled(
        shallow(
          <AccountForm
            error={new Error('Test error')}
            konnector={fixtures.konnector}
            initialValues={values}
            onSubmit={onSubmit}
            t={t}
          />,
          // Avoid componentDidMount
          { disableLifecycleMethods: true }
        )
      )
    })
  })

  it('should call onSubmit on click', () => {
    const wrapper = shallow(
      <AccountForm
        account={fixtures.account}
        konnector={fixtures.konnectorWithOptionalFields}
        onSubmit={onSubmit}
        t={t}
      />
    )
    wrapper
      .dive()
      .find('DefaultButton')
      .simulate('click')

    expect(onSubmit).toHaveBeenCalled()
  })

  describe('focusNext', () => {
    const loginInput = document.createElement('input')
    const passwordInput = document.createElement('input')

    it('should focus next input', () => {
      const wrapper = shallow(
        <AccountForm konnector={fixtures.konnector} onSubmit={onSubmit} t={t} />
      )

      wrapper.instance().inputs = { login: loginInput, password: passwordInput }
      wrapper.instance().inputs.login.focus()
      wrapper.instance().inputFocused = loginInput

      const firstFocus = wrapper.instance().focusNext()
      expect(firstFocus).toEqual(passwordInput)

      wrapper.instance().inputFocused = passwordInput
      const secondFocus = wrapper.instance().focusNext()
      expect(secondFocus).toBeNull()
    })
  })

  describe('handleKeyUp', () => {
    it('should ignore other keys than ENTER', () => {
      isMobile.mockReturnValue(false)
      const wrapper = shallow(
        <AccountForm
          konnector={fixtures.konnector}
          onSubmit={onSubmit}
          t={t}
        />,
        // Avoid componentDidMount
        { disableLifecycleMethods: true }
      )

      wrapper.instance().handleSubmit = jest.fn()

      wrapper.instance().handleKeyUp({ key: 'Space' }, {})
      expect(wrapper.instance().handleSubmit).not.toHaveBeenCalled()
    })

    it('should submit form', () => {
      isMobile.mockReturnValue(false)
      const wrapper = shallow(
        <AccountForm
          konnector={fixtures.konnector}
          onSubmit={onSubmit}
          t={t}
        />,
        // Avoid componentDidMount
        { disableLifecycleMethods: true }
      )

      wrapper.instance().handleSubmit = jest.fn()
      wrapper.instance().isSubmittable = jest.fn().mockReturnValue(true)

      wrapper.instance().handleKeyUp({ key: 'Enter' }, {})
      expect(wrapper.instance().handleSubmit).toHaveBeenCalled()
    })

    it('should not submit form', () => {
      isMobile.mockReturnValue(false)
      const wrapper = shallow(
        <AccountForm
          konnector={fixtures.konnector}
          onSubmit={onSubmit}
          t={t}
        />,
        // Avoid componentDidMount
        { disableLifecycleMethods: true }
      )

      wrapper.instance().handleSubmit = jest.fn()
      wrapper.instance().isSubmittable = jest.fn().mockReturnValue(false)

      wrapper.instance().handleKeyUp({ key: 'Enter' }, {})
      expect(wrapper.instance().handleSubmit).not.toHaveBeenCalled()
    })

    it('should focus next input on mobile', () => {
      isMobile.mockReturnValue(true)
      const wrapper = shallow(
        <AccountForm
          konnector={fixtures.konnector}
          onSubmit={onSubmit}
          t={t}
        />,
        // Avoid componentDidMount
        { disableLifecycleMethods: true }
      )

      wrapper.instance().focusNext = jest
        .fn()
        .mockReturnValue(document.createElement('input'))

      wrapper.instance().handleKeyUp({ key: 'Enter' }, {})
      expect(wrapper.instance().focusNext).toHaveBeenCalled()
    })

    it('should submit form on mobile', () => {
      isMobile.mockReturnValue(true)
      const wrapper = shallow(
        <AccountForm
          konnector={fixtures.konnector}
          onSubmit={onSubmit}
          t={t}
        />,
        // Avoid componentDidMount
        { disableLifecycleMethods: true }
      )

      wrapper.instance().focusNext = jest.fn().mockReturnValue(null)
      wrapper.instance().isSubmittable = jest.fn().mockReturnValue(true)
      wrapper.instance().handleSubmit = jest.fn()

      wrapper.instance().handleKeyUp({ key: 'Enter' }, {})
      expect(wrapper.instance().focusNext).toHaveBeenCalled()
      expect(wrapper.instance().handleSubmit).toHaveBeenCalled()
    })

    it('should not submit form on mobile', () => {
      isMobile.mockReturnValue(true)
      const wrapper = shallow(
        <AccountForm
          konnector={fixtures.konnector}
          onSubmit={onSubmit}
          t={t}
        />,
        // Avoid componentDidMount
        { disableLifecycleMethods: true }
      )

      wrapper.instance().focusNext = jest.fn().mockReturnValue(null)
      wrapper.instance().isSubmittable = jest.fn().mockReturnValue(false)
      wrapper.instance().handleSubmit = jest.fn()

      wrapper.instance().handleKeyUp({ key: 'Enter' }, {})
      expect(wrapper.instance().focusNext).toHaveBeenCalled()
      expect(wrapper.instance().handleSubmit).not.toHaveBeenCalled()
    })
  })

  describe('with read-only identifier', () => {
    it('should render a read-only identifier field if the account has a relationship with a vault cipher', () => {
      const wrapper = mount(
        <AccountForm
          konnector={fixtures.konnector}
          onSubmit={onSubmit}
          t={t}
          account={{
            ...fixtures.account,
            relationships: {
              vaultCipher: {
                _id: 'fake-cipher-id',
                _type: 'com.bitwarden.ciphers',
                _protocol: 'bitwarden'
              }
            }
          }}
        />,
        {
          context: { t },
          childContextTypes: {
            t: PropTypes.func
          }
        }
      )

      const hiddenInput = wrapper.find('input[type="hidden"][name="username"]')

      expect(hiddenInput).toHaveLength(1)
    })
  })
})
