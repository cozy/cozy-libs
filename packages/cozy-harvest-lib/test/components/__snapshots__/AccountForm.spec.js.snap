// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountForm should not render error 1`] = `
<div
  onFocusCapture={[Function]}
  onKeyUp={[Function]}
>
  <AccountFields
    container={null}
    fields={
      Object {
        "passphrase": Object {
          "encrypted": true,
          "required": true,
          "type": "password",
        },
        "username": Object {
          "encrypted": false,
          "required": true,
          "role": "identifier",
          "type": "text",
        },
      }
    }
    hasError={false}
    initialValues={Object {}}
    inputRefByName={[Function]}
    t={
      [MockFunction] {
        "calls": Array [
          Array [
            "accountForm.submit.label",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": "accountForm.submit.label",
          },
        ],
      }
    }
  />
  <DefaultButton
    className="u-mt-2 u-mb-1-half"
    disabled={false}
    extension="full"
    label="accountForm.submit.label"
    onClick={[Function]}
  />
</div>
`;

exports[`AccountForm should render 1`] = `
<div
  onFocusCapture={[Function]}
  onKeyUp={[Function]}
>
  <AccountFields
    container={null}
    fields={
      Object {
        "passphrase": Object {
          "encrypted": true,
          "required": true,
          "type": "password",
        },
        "username": Object {
          "encrypted": false,
          "required": true,
          "role": "identifier",
          "type": "text",
        },
      }
    }
    initialValues={Object {}}
    inputRefByName={[Function]}
    t={
      [MockFunction] {
        "calls": Array [
          Array [
            "accountForm.submit.label",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": "accountForm.submit.label",
          },
        ],
      }
    }
  />
  <DefaultButton
    className="u-mt-2 u-mb-1-half"
    disabled={true}
    extension="full"
    label="accountForm.submit.label"
    onClick={[Function]}
  />
</div>
`;

exports[`AccountForm should render error 1`] = `
<div
  onFocusCapture={[Function]}
  onKeyUp={[Function]}
>
  <Wrapper
    className="u-mb-1"
    error={[Error: Test error]}
    konnector={
      Object {
        "fields": Object {
          "passphrase": Object {
            "type": "password",
          },
          "username": Object {
            "type": "text",
          },
        },
      }
    }
  />
  <AccountFields
    container={null}
    fields={
      Object {
        "passphrase": Object {
          "encrypted": true,
          "required": true,
          "type": "password",
        },
        "username": Object {
          "encrypted": false,
          "required": true,
          "role": "identifier",
          "type": "text",
        },
      }
    }
    hasError={false}
    initialValues={Object {}}
    inputRefByName={[Function]}
    t={
      [MockFunction] {
        "calls": Array [
          Array [
            "accountForm.submit.label",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": "accountForm.submit.label",
          },
        ],
      }
    }
  />
  <DefaultButton
    className="u-mt-2 u-mb-1-half"
    disabled={false}
    extension="full"
    label="accountForm.submit.label"
    onClick={[Function]}
  />
</div>
`;
